tests/Models/OAuth/TokenTest.php

```diff
@@ -12,47 +12,139 @@
 use App\Models\User;
 use Database\Factories\OAuth\RefreshTokenFactory;
 use Illuminate\Support\Facades\Event;
-use Laravel\Passport\Passport;
 use Tests\TestCase;
 
 class TokenTest extends TestCase
 {
-    public function testAuthCodeChatWriteAllowsSelf()
+    public static function dataProviderForTestAuthCodeChatScopesAllowsSelf()
+    {
+        return static::chatScopes()->map(fn ($scope) => [$scope]);
+    }
+
+    public static function dataProviderForTestAuthCodeChatScopesRequiresBotGroup()
+    {
+        $data = [];
+        foreach (static::chatScopes() as $scope) {
+            $data[] = [$scope, null, true];
+            $data[] = [$scope, 'admin', true];
+            $data[] = [$scope, 'bng', true];
+            $data[] = [$scope, 'bot', false];
+            $data[] = [$scope, 'gmt', true];
+            $data[] = [$scope, 'nat', true];
+        }
+
+        return $data;
+    }
+
+    public static function dataProviderForTestDelegationNotAllowedScopes()
+    {
+        $delegationExceptDelegate = array_diff(Token::SCOPES_REQUIRE_DELEGATION, ['delegate']);
+
+        $data = [];
+        foreach (static::allPassportScopeIds() as $scope) {
+            match (true) {
+                $scope === 'delegate' => null,
+                in_array($scope, $delegationExceptDelegate, true) => $data[] = [[$scope], 'delegate_required'],
+                !in_array($scope, Token::SCOPES_REQUIRE_DELEGATION, true) => $data[] = [[$scope], null],
+                default => throw new \Exception("{$scope} was not handled."),
+            };
+        }
+
+        return $data;
+    }
+
+    public static function dataProviderForTestDelegationRequiredScopes()
+    {
+        $delegationExceptDelegate = array_diff(Token::SCOPES_REQUIRE_DELEGATION, ['delegate']);
+
+        $data = [];
+        foreach (static::allPassportScopeIds() as $scope) {
+            match (true) {
+                $scope === 'delegate' => $data[] = [['delegate'], null],
+                in_array($scope, $delegationExceptDelegate, true) => $data[] = [[$scope, 'delegate'], null],
+                !in_array($scope, Token::SCOPES_REQUIRE_DELEGATION, true) => $data[] = [[$scope, 'delegate'], 'delegate_invalid_combination'],
+                default => throw new \Exception("{$scope} was not handled."),
+            };
+        }
+
+        return $data;
```

Comment:
can be `return array_map()`? (and for the other one maybe with additional `array_reject_null`)